File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.cc
#include <bits/stdc++.h>
using namespace std;

int n;
vector<int> adj[100001];
int p[100001];

void dfs(int cur){
  for(int nxt:adj[cur]){
    if(p[cur] == nxt) continue;

    p[nxt] = cur;
    dfs(nxt);
  }
}

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> n;

  for(int i=0; i<n-1; i++){
    int a, b;
    cin >> a >> b;

    adj[a].push_back(b);
    adj[b].push_back(a);
  }

  dfs(1);

  for(int i=2; i<=n; i++){
    cout << p[i] << '\n';
  }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

  public static int n;
  public static List<Integer>[] adj = new ArrayList[100002];
  public static int[] p = new int[100002];

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = null;

    adj = new ArrayList[100002];
    for(int i=0; i<100002; i++){
      adj[i] = new ArrayList<>();
    }

    n = Integer.parseInt(br.readLine());
    for(int i=0; i<n-1; i++){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());

      adj[u].add(v);
      adj[v].add(u);
    }

    dfs(1);
    for(int i=2; i<=n; i++){
      System.out.println(p[i]);
    }
  }

  public static void dfs(int cur){
    for(int nxt:adj[cur]){
      if(p[cur] == nxt) continue;

      p[nxt] = cur;
      dfs(nxt);
    }
  }
}
